lib/rust/ensogl/component/grid-view/src/lib.rs

```diff
@@ -0,0 +1,508 @@
+    }
+}
+
+impl<E: Entry> Model<E, E::Params> {
+    fn update_entry(
+        &self,
+        row: Row,
+        col: Col,
+        model: E::Model,
+        entry_size: Vector2,
+        text_layer: &Option<WeakLayer>,
+    ) {
+        use std::collections::hash_map::Entry::*;
+        let mut visible_entries = self.visible_entries.borrow_mut();
+        let mut free_entries = self.free_entries.borrow_mut();
+        let create_new_entry = || {
+            let text_layer = text_layer.as_ref().and_then(|l| l.upgrade());
+            self.entry_creation_ctx.create_entry(&text_layer)
+        };
+        let entry = match visible_entries.entry((row, col)) {
+            Occupied(entry) => entry.into_mut(),
+            Vacant(lack_of_entry) => {
+                let new_entry = free_entries.pop().unwrap_or_else(create_new_entry);
+                set_entry_position(&new_entry, row, col, entry_size);
+                self.display_object.add_child(&new_entry);
+                lack_of_entry.insert(new_entry)
+            }
+        };
+        entry.frp().set_model(model);
+    }
+}
+
+
+
+// ================
+// === GridView ===
+// ================
+
+/// A template for [`GridView`] structure, where entry parameters and model are separate generic
+/// arguments.
+///
+/// It may be useful when using GridView in parametrized structs, where we want to avoid rewriting
+/// `Entry` bound in each place. Otherwise it's better to use [`GridView`].
+///
+/// Please note that some bounds are still required, as we use [`Widget`] and [`Frp`] nodes.
+#[derive(CloneRef, Debug, Deref, Derivative)]
+#[derivative(Clone(bound = ""))]
+pub struct GridViewTemplate<E: 'static, M: frp::node::Data, P: frp::node::Data> {
+    widget: Widget<Model<E, P>, Frp<M, P>>,
+}
+
+/// Grid View Component.
+///
+/// This Component displays any kind of entry `E` in a grid. To have it working, you need to
+/// * Set entries size ([`Frp::set_entries_size`]),
+/// * Declare (and keep up-to-date) the visible area ([`Frp::set_viewport`]),
+/// * Set up logic for providing models (see _Requesting for Models_ section).
+/// * Optionally entries parameters, if given entry does not have sensible default.
+/// * Finally reset the content, providing number of rows and columns ([`Frp::reset_entries`]).
+///
+/// # Positioning
+///
+/// Please mark, that this structure has its left-top corner docked to (0, 0) point of parent
+/// display object, as this is more intuitive way with handling grids.
+///
+/// # Entries Instantiation
+///
+/// The entry should implement [`Entry`] trait. Entries are instantiated lazily, only those visible
+/// in provided [`Frp::view_area`]. Once entries are no longer visible, are detached, but not
+/// dropped and may be re-used to display new entries when needed. This way we can achieve very
+/// efficient scrolling.
+///
+/// ## Requesting for Models
+///
+/// Once an entry is uncovered, the Grid View emits [`Frp::model_for_entry_needed`]. Then the proper
```

Comment:
What with spinners? How can we provide the GridView with a spinner implementation?

Reply:
You can have:
* An Entry Model being an enum with "Loading" and "Loaded" variants
* An Entry which will display spinner or content depending on situation
I will add to the doc info that "model_for_entry" may be used to update already visible entries as well.